crates/bevy_ecs/src/query/access.rs

```diff
@@ -325,6 +318,84 @@ impl<T: SparseSetIndex> FilteredAccess<T> {
     }
 }
 
+// A struct to express something like `Or<(With<A>, With<B>)>`.
+// Filters like `(With<A>, Or<(With<B>, With<C>)>` are expanded into `Or<((With<A>, With<B>), (With<A>, With<C>))>`.
+#[derive(Clone, Eq, PartialEq)]
+struct ExpandedOrAccess<T, I> {
+    arr: SmallVec<[FixedBitSet; 8]>,
+    // To prevent mixing `With` and `Without` filters.
+    _filter_type: PhantomData<T>,
+    _index_type: PhantomData<I>,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+struct WithMarker;
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+struct WithoutMarker;
+
+impl<T, I: SparseSetIndex + fmt::Debug> fmt::Debug for ExpandedOrAccess<T, I> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_list()
+            .entries(self.arr.iter().map(FormattedBitSet::<I>::new))
+            .finish()
+    }
+}
+
+impl<T, I: SparseSetIndex> Default for ExpandedOrAccess<T, I> {
+    fn default() -> Self {
+        Self {
+            arr: smallvec::smallvec![FixedBitSet::default()],
+            _filter_type: PhantomData,
+            _index_type: PhantomData,
+        }
+    }
+}
+
+impl<T, I: SparseSetIndex> ExpandedOrAccess<T, I> {
+    fn add(&mut self, index: usize) {
+        for with in &mut self.arr {
+            with.grow(index + 1);
+            with.insert(index);
+        }
+    }
+
+    fn extend_with_or(&mut self, other: &Self) {
+        self.arr.append(&mut other.arr.clone());
+    }
+
+    fn union_with(&mut self, other: &Self) {
+        if other.arr.len() == 1 {
+            for with in &mut self.arr {
+                with.union_with(&other.arr[0]);
+            }
+            return;
+        }
+
+        let mut new_with = SmallVec::with_capacity(self.arr.len() * other.arr.len());
+        for with in &self.arr {
+            for other_with in &other.arr {
+                let mut w = with.clone();
+                w.union_with(other_with);
+                new_with.push(w);
+            }
+        }
+        self.arr = new_with;
+    }
+}
+
+impl<I: SparseSetIndex> ExpandedOrAccess<WithMarker, I> {
+    fn is_disjoint(&self, other: &ExpandedOrAccess<WithoutMarker, I>) -> bool {
+        // We need to make sure that every `without` variant is disjoint with any `with`.
```

Comment:
This statement isn't obviously correct: can you add more justification or otherwise improve clarity?

Reply:
(marking as resolved as the implementation and comments changed significantly)